domains/client/domain-operator/src/operator.rs

```diff
@@ -178,19 +191,167 @@ where
             domain_block_processor.clone(),
         );
 
+        let snap_sync_orchestrator = params
+            .consensus_chain_sync_params
+            .as_ref()
+            .map(|p| p.snap_sync_orchestrator.clone());
+
+        let sync_params = params
+            .consensus_chain_sync_params
+            .map(|consensus_sync_params| SyncParams {
+                domain_client: params.client.clone(),
+                domain_network_service_handle: params.domain_network_service_handle,
+                sync_service: params.sync_service,
+                consensus_client: params.consensus_client.clone(),
+                domain_block_downloader: params.block_downloader.clone(),
+                consensus_chain_sync_params: consensus_sync_params,
+                domain_fork_id: params.domain_fork_id,
+                receipt_provider: params.domain_execution_receipt_provider,
+            });
+
+        if let Some(sync_params) = sync_params {
+            let domain_sync_task = {
+                async move {
+                    let info = sync_params.domain_client.info();
+                    // Only attempt snap sync with genesis state
+                    // TODO: Support snap sync from any state once
+                    //  https://github.com/paritytech/polkadot-sdk/issues/5366 is resolved
+                    if info.best_hash == info.genesis_hash {
+                        info!("Starting domain snap sync...");
+
+                        let result = snap_sync(sync_params).await;
+
+                        match result {
+                            Ok(_) => {
+                                info!("Domain snap sync completed.");
+                            }
+                            Err(err) => {
+                                error!(%err, "Domain snap sync failed.");
+
+                                // essential task failed
+                                return;
+                            }
+                        };
+                    } else {
+                        debug!("Snap sync can only work with genesis state, skipping");
+
+                        // essential task failed
+                        return;
+                    }
+
+                    // Don't exit essential task.
+                    pending().await
```

Comment:
This line prevents the essential task from exiting, which will cause the process to exit.

Reply:
Yes, I read the comment 🙂

When we implement snap sync from any height, will this `pending` be replaced with an infinite loop that waits to see if there’s a large height gap?

(If not, then it might be possible to combine this task with the other tasks that wait on it finishing.)

Reply:
Substrate code seems to evolve in a direction that will eventually allow us to snap sync repeatedly (It is not the case now).  But even if it is not I was asked to move it to the separate task in the previous iterations of the review.